mapreduce/src/main/java/uk/gov/gchq/palisade/PalisadeInputFormat.java

```diff
@@ -0,0 +1,391 @@
+
+    /**
+     * Default number of mappers to use. Hint only. Zero means unlimited.
+     */
+    public static final int DEFAULT_MAX_MAP_HINT = 0;
+
+    /**
+     * The Palisade service instance that serves as the central point for gathering data.
+     */
+    private static PalisadeService palisadeService;
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * This method will try to split the resource details returned from the Palisade service across multiple input
+     * splits as guided by the maxmimum mappers hint. However, we can only honour this hint so far, for example, each
+     * separate data request will generate at least one mapper. At the moment we don't try to balance input splits
+     * across multiple data requests. This may be added in the future.
+     *
+     * @throws IllegalStateException if no data requests have been added with {@link PalisadeInputFormat#addDataRequest(JobContext,
+     *                               RegisterDataRequest)} or if no Palisade service has been set or if the maximum map
+     *                               hint is negative
+     * @see PalisadeInputFormat#setMaxMapTasksHint(JobContext, int)
+     */
+    @Override
+    public List<InputSplit> getSplits(final JobContext context) throws IOException {
+        Objects.requireNonNull(context, "context");
+        //check we have a service set
+        if (palisadeService == null) {
+            throw new IllegalStateException("no Palisade service has been specified");
+        }
+        //get the list for this job
+        List<RegisterDataRequest> reqs = getDataRequests(context);
+        //sanity check
+        if (reqs.isEmpty()) {
+            throw new IllegalStateException("No data requests have been registered for this job");
+        }
+        //how many mappers hinted at?
+        int maxMapHint = getMaxMapTasksHint(context);
+        if (maxMapHint < 0) {
+            throw new IllegalStateException("Max map hint illegally set to negative number");
+        }
+
+        //store local and call through method, don't access direct!
+        PalisadeService serv = getPalisadeService();
+
+        ConcurrentLinkedDeque<InputSplit> splits = new ConcurrentLinkedDeque<>();
+        /*Each RegisterDataRequest may result in multiple resources, each of which should be in it's own input split.
+        *These may complete at different rates in any order, so we need to know when all of them have finished to know
+        * that all the resources have been added to the final list. Therefore, we use a Phaser to keep count of the number
+        * of registration requests that have been completed.*/
+        final Phaser counter = new Phaser(1); //register self
+        final int maxCounter = (maxMapHint == 0) ? Integer.MAX_VALUE : maxMapHint;
+        //create a stream for round robining resources
+
+        for (RegisterDataRequest req : reqs) {
```

Comment:
Rather than using the Phaser you could just collect all the futures then use CompletableFuture.allOf(futures).join() to block until they have all completed.
e.g:
```java
final CompletableFuture[] futures = new CompletableFuture[reqs.size()];
int i = 0;
for (RegisterDataRequest req : reqs) {
    //this iterator determines which mapper gets which resource
    final PrimitiveIterator.OfInt indexIt = IntStream.iterate(0, x -> (x + 1) % maxCounter).iterator();
    //send request to the Palisade service
    futures[i] = serv.registerDataRequest(req)
            //when the response comes back create input splits based on max mapper hint
            .thenApplyAsync(response -> PalisadeInputFormat.toInputSplits(response, indexIt))
                    //then add them to the master list (which is thread safe)
            .thenAccept(splits::addAll);
    i++;
}
CompletableFuture.allOf(futures).join();
```

Reply:
Done